store/sqlstore/thread_store.go

```diff
@@ -49,6 +54,11 @@ func threadToSlice(thread *model.Thread) []interface{} {
 func (s *SqlThreadStore) createIndexesIfNotExists() {
 	s.CreateIndexIfNotExists("idx_threads_last_reply_at", "Threads", "LastReplyAt")
 	s.CreateIndexIfNotExists("idx_threads_post_id", "Threads", "PostId")
+
+	s.CreateIndexIfNotExists("idx_thread_memberships_last_update_at", "ThreadMemberships", "LastUpdated")
+	s.CreateIndexIfNotExists("idx_thread_memberships_last_view_at", "ThreadMemberships", "LastViewed")
+	s.CreateIndexIfNotExists("idx_thread_memberships_post_id", "ThreadMemberships", "PostId")
```

Comment:
The `idx_thread_memberships_post_id` index is redundant because the primary key automatically creates a multi-column index for which `PostId` is the first column. The DBs should use that index for `PostId` lookups, so creating this index uses unnecessary storage and slows backups, restores, and writes.

Reply:
@mkraft what about UserId?

Reply:
I think that's still necessary because it's not the first position of the primary key index. But I would verify it.

Reply:
This shows that the query plan is almost identical with the difference just in the index used:

```
mattermost_test=# explain (format json) select * from ThreadMemberships where PostId = 'foo';
                        QUERY PLAN
-----------------------------------------------------------
 [                                                        +
   {                                                      +
     "Plan": {                                            +
       "Node Type": "Bitmap Heap Scan",                   +
       "Relation Name": "threadmemberships",              +
       "Alias": "threadmemberships",                      +
       "Startup Cost": 4.16,                              +
       "Total Cost": 9.50,                                +
       "Plan Rows": 2,                                    +
       "Plan Width": 157,                                 +
       "Recheck Cond": "((postid)::text = 'foo'::text)",  +
       "Plans": [                                         +
         {                                                +
           "Node Type": "Bitmap Index Scan",              +
           "Parent Relationship": "Outer",                +
           "Index Name": "idx_thread_memberships_post_id",+
           "Startup Cost": 0.00,                          +
           "Total Cost": 4.16,                            +
           "Plan Rows": 2,                                +
           "Plan Width": 0,                               +
           "Index Cond": "((postid)::text = 'foo'::text)" +
         }                                                +
       ]                                                  +
     }                                                    +
   }                                                      +
 ]
(1 row)
```
```
mattermost_test=# drop index idx_thread_memberships_post_id;
DROP INDEX
```
```
mattermost_test=# explain (format json) select * from ThreadMemberships where PostId = 'foo';
                        QUERY PLAN
----------------------------------------------------------
 [                                                       +
   {                                                     +
     "Plan": {                                           +
       "Node Type": "Bitmap Heap Scan",                  +
       "Relation Name": "threadmemberships",             +
       "Alias": "threadmemberships",                     +
       "Startup Cost": 4.16,                             +
       "Total Cost": 9.50,                               +
       "Plan Rows": 2,                                   +
       "Plan Width": 157,                                +
       "Recheck Cond": "((postid)::text = 'foo'::text)", +
       "Plans": [                                        +
         {                                               +
           "Node Type": "Bitmap Index Scan",             +
           "Parent Relationship": "Outer",               +
           "Index Name": "threadmemberships_pkey",       +
           "Startup Cost": 0.00,                         +
           "Total Cost": 4.16,                           +
           "Plan Rows": 2,                               +
           "Plan Width": 0,                              +
           "Index Cond": "((postid)::text = 'foo'::text)"+
         }                                               +
       ]                                                 +
     }                                                   +
   }                                                     +
 ]
(1 row)
```

Reply:
And here's the same analysis done in MySQL:
```
mysql> explain format=json select * from ThreadMemberships where PostId = 'foo';
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "table": {
      "table_name": "ThreadMemberships",
      "access_type": "ref",
      "possible_keys": [
        "PRIMARY",
        "idx_thread_memberships_post_id"
      ],
      "key": "PRIMARY",
      "used_key_parts": [
        "PostId"
      ],
      "key_length": "106",
      "ref": [
        "const"
      ],
      "rows": 1,
      "filtered": 100,
      "attached_condition": "(`mattermost_test`.`ThreadMemberships`.`PostId` = 'foo')"
    }
  }
}
1 row in set, 1 warning (0.00 sec)
```
```
mysql> drop index idx_thread_memberships_post_id on ThreadMemberships;
Query OK, 0 rows affected (0.01 sec)
Records: 0  Duplicates: 0  Warnings: 0
```
```
mysql> explain format=json select * from ThreadMemberships where PostId = 'foo';
*************************** 1. row ***************************
EXPLAIN: {
  "query_block": {
    "select_id": 1,
    "table": {
      "table_name": "ThreadMemberships",
      "access_type": "ref",
      "possible_keys": [
        "PRIMARY"
      ],
      "key": "PRIMARY",
      "used_key_parts": [
        "PostId"
      ],
      "key_length": "106",
      "ref": [
        "const"
      ],
      "rows": 1,
      "filtered": 100,
      "attached_condition": "(`mattermost_test`.`ThreadMemberships`.`PostId` = 'foo')"
    }
  }
}
1 row in set, 1 warning (0.00 sec)
```

Reply:
thanks for the details, @mkraft . I'm going to remove the first one and leave `idx_thread_memberships_user_id` for now.